include/GafferScene/ClosurePlug.h

```diff
@@ -0,0 +1,77 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2017, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENE_CLOSUREPLUG_H
+#define GAFFERSCENE_CLOSUREPLUG_H
+
+#include "IECore/CompoundObject.h"
+
+#include "Gaffer/Plug.h"
+
+#include "GafferScene/TypeIds.h"
+
+namespace GafferScene
+{
+
+/// Plug that provides a proxy for representing closure types when
+/// loader a shader from OSL or a renderer.  We probably won't be able
+/// to set or get closure plugs, but we need to be able to connect 
+/// them, and they should only connect to other closure plugs.
+class ClosurePlug : public Gaffer::Plug
```

Comment:
I think I mentioned this in an earlier email, but yeah, this was the part I was most uncertain in.  It isn't any different than the OSL case, and if we decide to use a specific plug type here, then we should definitely use it for OSL as well ( I've now tested, and it is kosher in Arnold to connect Arnold closure -> OSL or OSL closure -> Arnold ).

As you point out, we have managed without it in OSL so far, so we definitely don't need it ( though I'm not sure if anyone has actually used OSL closures in Gaffer much yet ).  The main reason I went this way to start with is that I had no idea which was better and I needed to get something released for testing ASAP, and it seemed easier to do a compatibility config later to remove the custom plug type than to add it if we decided it was desirable.

Talking with Don, he thought that a custom plug type made things a lot clearer, and I agree.  It is a special type of data and closure plugs should only ever be allowed to connect to other closure plugs.  We can achieve this by using a default Plug type and putting a custom acceptsInput on any node that ever has a closure plug on it, but this seems like a complex and failure prone way to achieve the same thing as actually having a ClosurePlug type?

( Oh, and the shaderOutPlug was just a sloppy copy/paste while I was setting up this prototype ).

Reply:
Oh, and also, looking at the current logic in OSLShader acceptsInput, it currently allows you to connect any raw Plug into a closure Plug?  That seems like it could cause problems if anything in Gaffer uses Gaffer::Plug to mean anything other than a closure plug?